Unable to open files when folder has too much content

**Describe the bug**
Unable to image files through gallery

**To Reproduce**
Steps to reproduce the behavior:
Had over 8k images in a folder and try to open an image throwing a crash.

**Smartphone (please complete the following information):**
 - Device: Oneplus 6t
 - OS: Android 10
 - Rooted: No
 - Version: 3.5.3

**Additional context**
Add any other context about the problem here.

Stacktrace 

```
2020-10-08 20:54:52.768 28345-28345/com.amaze.filemanager.debug E/JavaBinder: !!! FAILED BINDER TRANSACTION !!!  (parcel size = 6622424)
2020-10-08 20:54:52.769 28345-28345/com.amaze.filemanager.debug W/ActivityStopInfo: Bundle stats:
2020-10-08 20:54:52.772 28345-28345/com.amaze.filemanager.debug W/ActivityStopInfo:   android:viewHierarchyState [size=7120]
2020-10-08 20:54:52.774 28345-28345/com.amaze.filemanager.debug W/ActivityStopInfo:     android:views [size=7072]
2020-10-08 20:54:52.775 28345-28345/com.amaze.filemanager.debug W/ActivityStopInfo:       0x7f09017e [size=2864]
2020-10-08 20:54:53.310 28345-28345/com.amaze.filemanager.debug W/ActivityStopInfo:   android:support:fragments [size=6614472]
2020-10-08 20:54:53.311 28345-28345/com.amaze.filemanager.debug W/ActivityStopInfo: PersistableBundle stats:
2020-10-08 20:54:53.311 28345-28345/com.amaze.filemanager.debug W/ActivityStopInfo:   [null]
2020-10-08 20:54:53.311 28345-28345/com.amaze.filemanager.debug D/AndroidRuntime: Shutting down VM
2020-10-08 20:54:53.316 28345-28345/com.amaze.filemanager.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.amaze.filemanager.debug, PID: 28345
    java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 6622424 bytes
        at android.app.servertransaction.PendingTransactionActions$StopInfo.run(PendingTransactionActions.java:161)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7643)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
     Caused by: android.os.TransactionTooLargeException: data parcel size 6622424 bytes
        at android.os.BinderProxy.transactNative(Native Method)
        at android.os.BinderProxy.transact(BinderProxy.java:550)
        at android.app.IActivityTaskManager$Stub$Proxy.activityStopped(IActivityTaskManager.java:4564)
        at android.app.servertransaction.PendingTransactionActions$StopInfo.run(PendingTransactionActions.java:145)
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7643) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950) 
2020-10-08 20:54:53.358 28345-28345/com.amaze.filemanager.debug I/Process: Sending signal. PID: 28345 SIG: 9

```

--- a/app/src/main/java/com/amaze/filemanager/filesystem/CustomFileObserver.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/CustomFileObserver.java
@@ -97,7 +97,7 @@ public class CustomFileObserver extends FileObserver {
     if (Build.VERSION.SDK_INT == Build.VERSION_CODES.M) {
       stopPollingSystem();
     } else {
-      super.startWatching();
+      super.stopWatching();
     }
   }
 

--- a/app/src/main/java/com/amaze/filemanager/filesystem/MediaStoreHack.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/MediaStoreHack.java
@@ -119,6 +119,12 @@ public class MediaStoreHack {
     return outputStream;
   }
 
+  /**
+   * Fallback to get uri from a path. Used only as a workaround for Kitkat ext SD card
+   * @param path file path
+   * @param context context
+   * @return uri of file
+   */
   public static Uri getUriFromFile(final String path, Context context) {
     ContentResolver resolver = context.getContentResolver();
 

--- a/app/src/main/java/com/amaze/filemanager/filesystem/MediaStoreHack.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/MediaStoreHack.java
@@ -121,6 +121,7 @@ public class MediaStoreHack {
 
   /**
    * Fallback to get uri from a path. Used only as a workaround for Kitkat ext SD card
+   *
    * @param path file path
    * @param context context
    * @return uri of file

--- a/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
@@ -197,9 +197,6 @@ public class MainActivity extends PermissionsActivity
   // This holds the Uris to be written at initFabToSave()
   private ArrayList<Uri> urisToBeSaved;
 
-  /** @deprecated use getCurrentMainFragment() */
-  public MainFragment mainFragment;
-
   public static final String PASTEHELPER_BUNDLE = "pasteHelper";
 
   private static final String KEY_DRAWER_SELECTED = "selectitem";
@@ -447,7 +444,6 @@ public class MainActivity extends PermissionsActivity
       oppathe1 = savedInstanceState.getString(KEY_OPERATED_ON_PATH);
       oparrayList = savedInstanceState.getParcelableArrayList(KEY_OPERATIONS_PATH_LIST);
       operation = savedInstanceState.getInt(KEY_OPERATION);
-      // mainFragment = (Main) savedInstanceState.getParcelable("main_fragment");
       int selectedStorage = savedInstanceState.getInt(KEY_DRAWER_SELECTED, 0);
       getDrawer().selectCorrectDrawerItem(selectedStorage);
     }
@@ -1219,9 +1215,6 @@ public class MainActivity extends PermissionsActivity
       outState.putParcelableArrayList(KEY_OPERATIONS_PATH_LIST, (oparrayList));
       outState.putInt(KEY_OPERATION, operation);
     }
-    /*if (mainFragment!=null) {
-        outState.putParcelable("main_fragment", mainFragment);
-    }*/
   }
 
   @Override
@@ -1926,25 +1919,25 @@ public class MainActivity extends PermissionsActivity
 
   @Override
   public void onPreExecute(String query) {
-    mainFragment.mSwipeRefreshLayout.setRefreshing(true);
-    mainFragment.onSearchPreExecute(query);
+    getCurrentMainFragment().mSwipeRefreshLayout.setRefreshing(true);
+    getCurrentMainFragment().onSearchPreExecute(query);
   }
 
   @Override
   public void onPostExecute(String query) {
-    mainFragment.onSearchCompleted(query);
-    mainFragment.mSwipeRefreshLayout.setRefreshing(false);
+    getCurrentMainFragment().onSearchCompleted(query);
+    getCurrentMainFragment().mSwipeRefreshLayout.setRefreshing(false);
   }
 
   @Override
   public void onProgressUpdate(HybridFileParcelable val, String query) {
-    mainFragment.addSearchResult(val, query);
+    getCurrentMainFragment().addSearchResult(val, query);
   }
 
   @Override
   public void onCancelled() {
-    mainFragment.reloadListElements(false, false, !mainFragment.IS_LIST);
-    mainFragment.mSwipeRefreshLayout.setRefreshing(false);
+    getCurrentMainFragment().reloadListElements(false, false, !getCurrentMainFragment().IS_LIST);
+    getCurrentMainFragment().mSwipeRefreshLayout.setRefreshing(false);
   }
 
   @Override

--- a/app/src/main/java/com/amaze/filemanager/ui/dialogs/GeneralDialogCreation.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/dialogs/GeneralDialogCreation.java
@@ -590,7 +590,7 @@ public class GeneralDialogCreation {
     }
 
     if (!forStorage && showPermissions) {
-      final MainFragment main = ((MainActivity) base).mainFragment;
+      final MainFragment main = ((MainActivity) base).getCurrentMainFragment();
       AppCompatButton appCompatButton = v.findViewById(R.id.permissionsButton);
       appCompatButton.setAllCaps(true);
 

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
@@ -382,7 +382,7 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
     reloadListElements(false, results, isPathLayoutGrid);
   }
 
-  void loadViews() {
+  private void loadViews() {
     if (CURRENT_PATH != null) {
       getMainActivity()
           .getAppbar()

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
@@ -384,17 +384,6 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
 
   private void loadViews() {
     if (CURRENT_PATH != null) {
-      getMainActivity()
-          .getAppbar()
-          .getBottomBar()
-          .updatePath(
-              CURRENT_PATH,
-              results,
-              MainActivityHelper.SEARCH_TEXT,
-              openMode,
-              folder_count,
-              file_count,
-              this);
       if ((LIST_ELEMENTS == null || LIST_ELEMENTS.size() == 0) && !results) {
         loadlist(CURRENT_PATH, true, OpenMode.FILE);
       } else {

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
@@ -48,7 +48,6 @@ import com.amaze.filemanager.filesystem.CustomFileObserver;
 import com.amaze.filemanager.filesystem.FileUtil;
 import com.amaze.filemanager.filesystem.HybridFile;
 import com.amaze.filemanager.filesystem.HybridFileParcelable;
-import com.amaze.filemanager.filesystem.MediaStoreHack;
 import com.amaze.filemanager.filesystem.PasteHelper;
 import com.amaze.filemanager.filesystem.cloud.CloudUtil;
 import com.amaze.filemanager.filesystem.files.CryptUtil;
@@ -960,10 +959,10 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
 
     Uri mediaStoreUri = Utils.getUriForBaseFile(getActivity(), baseFile);
     Log.d(
-            getClass().getSimpleName(),
-            mediaStoreUri.toString()
-                    + "\t"
-                    + MimeTypes.getMimeType(baseFile.getPath(), baseFile.isDirectory()));
+        getClass().getSimpleName(),
+        mediaStoreUri.toString()
+            + "\t"
+            + MimeTypes.getMimeType(baseFile.getPath(), baseFile.isDirectory()));
     Intent intent = new Intent();
     intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
     intent.setAction(Intent.ACTION_SEND);

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
@@ -162,7 +162,6 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
   private RecyclerView listView;
   private UtilitiesProvider utilsProvider;
   private HashMap<String, Bundle> scrolls = new HashMap<>();
-  private MainFragment ma = this;
   private View rootView;
   private View actionModeView;
   private FastScroller fastScroller;
@@ -214,21 +213,10 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
     primaryTwoColor = getMainActivity().getCurrentColorPreference().primarySecondTab;
   }
 
-  public void stopAnimation() {
-    if ((!adapter.stoppedAnimation)) {
-      for (int j = 0; j < listView.getChildCount(); j++) {
-        View v = listView.getChildAt(j);
-        if (v != null) v.clearAnimation();
-      }
-    }
-    adapter.stoppedAnimation = true;
-  }
-
   @Override
   public View onCreateView(
       LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
     rootView = inflater.inflate(R.layout.main_frag, container, false);
-    setRetainInstance(true);
     listView = rootView.findViewById(R.id.listView);
     mToolbarContainer = getMainActivity().getAppbar().getAppbarLayout();
     fastScroller = rootView.findViewById(R.id.fastscroll);
@@ -268,10 +256,8 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
     super.onActivityCreated(savedInstanceState);
 
     setHasOptionsMenu(false);
-    // getMainActivity() = (MainActivity) getActivity();
     initNoFileLayout();
     getSortModes();
-    this.setRetainInstance(false);
     HybridFile f = new HybridFile(OpenMode.UNKNOWN, CURRENT_PATH);
     f.generateMode(getActivity());
     getMainActivity().getAppbar().getBottomBar().setClickListener();
@@ -316,8 +302,9 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
                   if (columns == 0 || columns == -1) columns = 3;
                   if (!IS_LIST) mLayoutManagerGrid.setSpanCount(columns);
                 }
-                if (savedInstanceState != null && !IS_LIST)
-                  onSavedInstanceState(savedInstanceState);
+                if (!IS_LIST) {
+                  loadViews();
+                }
                 if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN) {
                   mToolbarContainer.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                 } else {
@@ -325,12 +312,17 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
                 }
               }
             });
+    loadViews();
+  }
 
-    if (savedInstanceState == null) {
-      loadlist(CURRENT_PATH, false, openMode);
-    } else {
-      if (IS_LIST) onSavedInstanceState(savedInstanceState);
+  public void stopAnimation() {
+    if ((!adapter.stoppedAnimation)) {
+      for (int j = 0; j < listView.getChildCount(); j++) {
+        View v = listView.getChildAt(j);
+        if (v != null) v.clearAnimation();
+      }
     }
+    adapter.stoppedAnimation = true;
   }
 
   void setGridLayoutSpanSizeLookup(GridLayoutManager mLayoutManagerGrid) {
@@ -390,60 +382,8 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
     reloadListElements(false, results, isPathLayoutGrid);
   }
 
-  @Override
-  public void onSaveInstanceState(Bundle outState) {
-    super.onSaveInstanceState(outState);
-
-    int index;
-    View vi;
-    if (listView != null) {
-      if (IS_LIST) {
-        index = (mLayoutManager).findFirstVisibleItemPosition();
-        vi = listView.getChildAt(0);
-      } else {
-        index = (mLayoutManagerGrid).findFirstVisibleItemPosition();
-        vi = listView.getChildAt(0);
-      }
-
-      int top = (vi == null) ? 0 : vi.getTop();
-
-      outState.putInt("index", index);
-      outState.putInt("top", top);
-      outState.putParcelableArrayList("list", LIST_ELEMENTS);
-      outState.putString("CURRENT_PATH", CURRENT_PATH);
-      outState.putBoolean("selection", selection);
-      outState.putInt("openMode", openMode.ordinal());
-      outState.putInt("folder_count", folder_count);
-      outState.putInt("file_count", file_count);
-
-      if (selection) {
-        outState.putIntegerArrayList("position", adapter.getCheckedItemsIndex());
-      }
-
-      outState.putBoolean("results", results);
-
-      if (openMode == OpenMode.SMB) {
-        outState.putString("SmbPath", smbPath);
-      }
-    }
-  }
-
-  void onSavedInstanceState(final Bundle savedInstanceState) {
-    Bundle b = new Bundle();
-    String cur = savedInstanceState.getString("CURRENT_PATH");
-
-    if (cur != null) {
-      b.putInt("index", savedInstanceState.getInt("index"));
-      b.putInt("top", savedInstanceState.getInt("top"));
-      scrolls.put(cur, b);
-
-      openMode = OpenMode.getOpenMode(savedInstanceState.getInt("openMode", 0));
-      if (openMode == OpenMode.SMB) smbPath = savedInstanceState.getString("SmbPath");
-      LIST_ELEMENTS = savedInstanceState.getParcelableArrayList("list");
-      CURRENT_PATH = cur;
-      folder_count = savedInstanceState.getInt("folder_count", 0);
-      file_count = savedInstanceState.getInt("file_count", 0);
-      results = savedInstanceState.getBoolean("results");
+  void loadViews() {
+    if (CURRENT_PATH != null) {
       getMainActivity()
           .getAppbar()
           .getBottomBar()
@@ -456,15 +396,17 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
               file_count,
               this);
       if ((LIST_ELEMENTS == null || LIST_ELEMENTS.size() == 0) && !results) {
-        loadlist(home, true, OpenMode.FILE);
+        loadlist(CURRENT_PATH, true, OpenMode.FILE);
       } else {
         reloadListElements(true, results, !IS_LIST);
       }
-      if (savedInstanceState.getBoolean("selection")) {
-        for (Integer index : savedInstanceState.getIntegerArrayList("position")) {
+      if (selection) {
+        for (Integer index : adapter.getCheckedItemsIndex()) {
           adapter.toggleChecked(index, null);
         }
       }
+    } else {
+      loadlist(home, true, OpenMode.FILE);
     }
   }
 
@@ -819,7 +761,7 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
       };
 
   public void home() {
-    ma.loadlist((ma.home), false, OpenMode.FILE);
+    loadlist((home), false, OpenMode.FILE);
   }
 
   /**
@@ -891,7 +833,7 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
           EncryptDecryptUtils.decryptFile(
               getContext(),
               getMainActivity(),
-              ma,
+              this,
               openMode,
               e.generateBaseFile(),
               getActivity().getExternalCacheDir().getPath(),
@@ -999,9 +941,9 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
 
     loadFilesListTask =
         new LoadFilesListTask(
-            ma.getActivity(),
+            getActivity(),
             path,
-            ma,
+            this,
             openMode,
             getBoolean(PREFERENCE_SHOW_THUMB),
             getBoolean(PREFERENCE_SHOW_HIDDENFILES),
@@ -1103,12 +1045,12 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
         adapter =
             new RecyclerAdapter(
                 getMainActivity(),
-                ma,
+                this,
                 utilsProvider,
                 sharedPref,
                 listView,
                 LIST_ELEMENTS,
-                ma.getActivity());
+                getActivity());
       } else {
         adapter.setItems(listView, new ArrayList<>(LIST_ELEMENTS));
       }
@@ -1334,8 +1276,6 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
         if (MainActivityHelper.SEARCH_TEXT != null) {
 
           // starting the search query again :O
-          getMainActivity().mainFragment =
-              (MainFragment) getMainActivity().getTabFragment().getCurrentTabFragment();
           FragmentManager fm = getMainActivity().getSupportFragmentManager();
 
           // getting parent path to resume search from there

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
@@ -958,37 +958,26 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
 
     getMainActivity().mReturnIntent = false;
 
+    Uri mediaStoreUri = Utils.getUriForBaseFile(getActivity(), baseFile);
+    Log.d(
+            getClass().getSimpleName(),
+            mediaStoreUri.toString()
+                    + "\t"
+                    + MimeTypes.getMimeType(baseFile.getPath(), baseFile.isDirectory()));
     Intent intent = new Intent();
-    if (getMainActivity().mRingtonePickerIntent) {
+    intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
+    intent.setAction(Intent.ACTION_SEND);
 
-      Uri mediaStoreUri = MediaStoreHack.getUriFromFile(baseFile.getPath(), getActivity());
-      Log.d(
-          getClass().getSimpleName(),
-          mediaStoreUri.toString()
-              + "\t"
-              + MimeTypes.getMimeType(baseFile.getPath(), baseFile.isDirectory()));
+    if (getMainActivity().mRingtonePickerIntent) {
       intent.setDataAndType(
           mediaStoreUri, MimeTypes.getMimeType(baseFile.getPath(), baseFile.isDirectory()));
       intent.putExtra(RingtoneManager.EXTRA_RINGTONE_PICKED_URI, mediaStoreUri);
-      getActivity().setResult(FragmentActivity.RESULT_OK, intent);
-      getActivity().finish();
     } else {
-
       Log.d("pickup", "file");
-
-      Intent intentresult = new Intent();
-
-      Uri resultUri = Utils.getUriForBaseFile(getActivity(), baseFile);
-      intentresult.setAction(Intent.ACTION_SEND);
-      intentresult.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
-
-      if (resultUri != null)
-        intentresult.setDataAndType(resultUri, MimeTypes.getExtension(baseFile.getPath()));
-
-      getActivity().setResult(FragmentActivity.RESULT_OK, intentresult);
-      getActivity().finish();
-      // mode.finish();
+      intent.setDataAndType(mediaStoreUri, MimeTypes.getExtension(baseFile.getPath()));
     }
+    getActivity().setResult(FragmentActivity.RESULT_OK, intent);
+    getActivity().finish();
   }
 
   LoadFilesListTask loadFilesListTask;

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/TabFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/TabFragment.java
@@ -196,17 +196,7 @@ public class TabFragment extends Fragment implements ViewPager.OnPageChangeListe
       if (fragment instanceof MainFragment) {
         MainFragment m = (MainFragment) fragment;
         if (i - 1 == MainActivity.currentTab && i == pos) {
-          mainActivity
-              .getAppbar()
-              .getBottomBar()
-              .updatePath(
-                  m.getCurrentPath(),
-                  m.results,
-                  MainActivityHelper.SEARCH_TEXT,
-                  m.openMode,
-                  m.folder_count,
-                  m.file_count,
-                  m);
+          updateBottomBar(m);
           mainActivity.getDrawer().selectCorrectDrawerItemForPath(m.getCurrentPath());
         }
         if (m.openMode == OpenMode.FILE) {
@@ -284,17 +274,7 @@ public class TabFragment extends Fragment implements ViewPager.OnPageChangeListe
       MainFragment ma = (MainFragment) fragment;
       if (ma.getCurrentPath() != null) {
         mainActivity.getDrawer().selectCorrectDrawerItemForPath(ma.getCurrentPath());
-        mainActivity
-            .getAppbar()
-            .getBottomBar()
-            .updatePath(
-                ma.getCurrentPath(),
-                ma.results,
-                MainActivityHelper.SEARCH_TEXT,
-                ma.openMode,
-                ma.folder_count,
-                ma.file_count,
-                ma);
+        updateBottomBar(ma);
       }
     }
 
@@ -331,8 +311,8 @@ public class TabFragment extends Fragment implements ViewPager.OnPageChangeListe
     }
   }
 
-  private void addNewTab(int num, String path) {
-    addTab(new Tab(num, path, path), "");
+  private void addNewTab(int num, String path, boolean currentTab) {
+    addTab(new Tab(num, path, path), "", currentTab);
   }
 
   /**
@@ -357,8 +337,8 @@ public class TabFragment extends Fragment implements ViewPager.OnPageChangeListe
       String currentFirstTab = Utils.isNullOrEmpty(firstTabPath) ? "/" : firstTabPath;
       String currentSecondTab = Utils.isNullOrEmpty(secondTabPath) ? firstTabPath : secondTabPath;
       if (addTab) {
-        addNewTab(1, currentSecondTab);
-        addNewTab(2, currentFirstTab);
+        addNewTab(1, currentSecondTab, true);
+        addNewTab(2, currentFirstTab, false);
       }
       tabHandler.addTab(new Tab(1, currentSecondTab, currentSecondTab));
       tabHandler.addTab(new Tab(2, currentFirstTab, currentFirstTab));
@@ -369,23 +349,23 @@ public class TabFragment extends Fragment implements ViewPager.OnPageChangeListe
     } else {
       if (path != null && path.length() != 0) {
         if (MainActivity.currentTab == 0) {
-          addTab(tab1, path);
-          addTab(tab2, "");
+          addTab(tab1, path, true);
+          addTab(tab2, "", false);
         }
 
         if (MainActivity.currentTab == 1) {
-          addTab(tab1, "");
-          addTab(tab2, path);
+          addTab(tab1, "", false);
+          addTab(tab2, path, true);
         }
       } else {
-        addTab(tab1, "");
-        addTab(tab2, "");
+        addTab(tab1, "", true);
+        addTab(tab2, "", false);
       }
     }
   }
 
-  private void addTab(@NonNull Tab tab, String path) {
-    Fragment main = new MainFragment();
+  private void addTab(@NonNull Tab tab, String path, boolean currentTab) {
+    MainFragment main = new MainFragment();
     Bundle b = new Bundle();
 
     if (path != null && path.length() != 0) {
@@ -401,6 +381,9 @@ public class TabFragment extends Fragment implements ViewPager.OnPageChangeListe
     fragments.add(main);
     mSectionsPagerAdapter.notifyDataSetChanged();
     mViewPager.setOffscreenPageLimit(4);
+    if (currentTab) {
+      updateBottomBar(main);
+    }
   }
 
   public Fragment getCurrentTabFragment() {
@@ -426,5 +409,19 @@ public class TabFragment extends Fragment implements ViewPager.OnPageChangeListe
       circleDrawable1.setImageDrawable(new ColorCircleDrawable(accentColor));
       circleDrawable2.setImageDrawable(new ColorCircleDrawable(Color.GRAY));
     }
+  }
+
+  private void updateBottomBar(MainFragment mainFragment) {
+    mainActivity
+        .getAppbar()
+        .getBottomBar()
+        .updatePath(
+            mainFragment.getCurrentPath(),
+            mainFragment.results,
+            MainActivityHelper.SEARCH_TEXT,
+            mainFragment.openMode,
+            mainFragment.folder_count,
+            mainFragment.file_count,
+            mainFragment);
   }
 }

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/TabFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/TabFragment.java
@@ -173,8 +173,6 @@ public class TabFragment extends Fragment implements ViewPager.OnPageChangeListe
             ThemedActivity.skinTwo : ThemedActivity.skin));
     mainActivity.updateViews(colorDrawable);*/
 
-    mainActivity.mainFragment = (MainFragment) getCurrentTabFragment();
-
     return rootView;
   }
 

--- a/app/src/main/java/com/amaze/filemanager/utils/MainActivityHelper.java
+++ b/app/src/main/java/com/amaze/filemanager/utils/MainActivityHelper.java
@@ -671,8 +671,6 @@ public class MainActivityHelper {
     task.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, fpath);*/
     // ma.searchTask = task;
     SEARCH_TEXT = query;
-    mainActivity.mainFragment =
-        (MainFragment) mainActivity.getTabFragment().getCurrentTabFragment();
     FragmentManager fm = mainActivity.getSupportFragmentManager();
     SearchWorkerFragment fragment =
         (SearchWorkerFragment) fm.findFragmentByTag(MainActivity.TAG_ASYNC_HELPER);

